Extend SimpleCache to allow for unbounded growth

aaronwinter opened the pull request:
In this PR, we allow users to create a (SimpleCache) cache with no artificial upper bound on the number of elements that it can host. To achieve that, the user will specify a `size <= 0` before building the cache.

bluele commented on simple.go:
```diff
@@ -18,7 +18,11 @@ func newSimpleCache(cb *CacheBuilder) *SimpleCache {
 
 func (c *SimpleCache) init() {
-	c.items = make(map[interface{}]*simpleItem, c.size)
+	if c.size == -1 {
```
I want `size==0` to means unbounded, so could you change to `c.size <= 0` ?

aaronwinter replied:
👍

bluele commented on simple.go:
```diff
@@ -58,7 +62,7 @@ func (c *SimpleCache) set(key, value interface{}) (interface{}, error) {
 	} else {
 		// Verify size not exceeded
-		if len(c.items) >= c.size {
+		if (len(c.items) >= c.size) && c.size != -1 {
```
@aaronwinter 
ref: https://github.com/bluele/gcache/pull/38/files#r127405430

Could you change to `if c.size > 0 && len(c.items) >= c.size` ?

kogan69 commented on simple.go:
```diff
@@ -18,7 +18,11 @@ func newSimpleCache(cb *CacheBuilder) *SimpleCache {
 
 func (c *SimpleCache) init() {
-	c.items = make(map[interface{}]*simpleItem, c.size)
+	if c.size == -1 {
```
Hi,
When do you guys expect this to be merged?
This is exactly the change I've been waiting for...

aaronwinter replied:
@kogan69 Hi, my bad sort of forgot about it. I will complete the PR by Wednesday this week.

kogan69 commented:
Thank you so much!


Cheers,
LK


> On Oct 1, 2017, at 7:38 PM, Erwan <notifications@github.com> wrote:
> 
> @aaronwinter commented on this pull request.
> 
> In simple.go:
> 
> > @@ -18,7 +18,11 @@ func newSimpleCache(cb *CacheBuilder) *SimpleCache {
>  }
>  
>  func (c *SimpleCache) init() {
> -	c.items = make(map[interface{}]*simpleItem, c.size)
> +	if c.size == -1 {
> @kogan69 Hi, my bad sort of forgot about it. I will complete the PR by Wednesday this week.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

aaronwinter commented:
@kogan69 I fixed the PR. Let me add a test suite for it and then it should be good to go for review and merge.

aaronwinter commented:
Added the test case, I'd like to add another to test for `SetWithExpire` but turns out the fake clock object we introduced is only package-exposed. I'll do another PR once we get that figured out.

This one is ready for review @bluele

bluele commented:
Thanks, @aaronwinter !
LGTM.
